### ISSUE TITLE
Data Request

### ISSUE THREAD

**Issue by wsnow99** [bolandih/2023_Stress-PINN#1]
I am writing to inquire about the availability of the data. My current research focuses on stress field prediction based on deep learning, and your data would be invaluable for comparative analysis. Your assistance in this matter would be greatly appreciated.

**Comment by bibiooi** [comment]
Can you provide specific data codes? Thank you very much

**Comment by wsnow99** [comment]
Sorry for the late reply. I am utilizing PointNet to predict the stress field, and here is my specific data codes.

# read coordinates and stress data according to the file name
data_raw = []
data_root = "data1213/all/"

file_name = os.listdir(data_root)
file_name = sorted(file_name, key = lambda x: int(x[5:-4]))

min_point_num = 1e7
for i, name in enumerate(file_name):
    data = sio.loadmat(os.path.join(data_root, name))["data"].astype(np.float32)
    index = np.arange(data.shape[1])
    
    data = data[:, data[0, :, 2] >= 0, :]
    data = data[:, data[0, :, 0] >= 0.05, :]
    data_length = data.shape[1]
    if data_length < min_point_num:
        min_point_num = data_length
    print(name, data.shape)
    data_raw.append(data)

# all_data
data_agg = []
for data in data_raw:
    index = np.arange(data.shape[1])
    np.random.shuffle(index)
    data = data[:, index[:min_point_num], :]
    data_agg.append(data)
data_agg = np.stack(data_agg)
print(data_agg.shape) # (137, 20, 2610, 9), 20 means the load step

# split train, valid, and test dataset
index = np.arange(len(data_agg))
np.random.shuffle(index)
train_index = index[: int(0.8*len(index))]
valid_index = index[int(0.8*len(index)) : int(0.9*len(index))]
test_index  = index[int(0.9*len(index)) :]

load = np.linspace(0, 100, 21, endpoint=True, dtype=np.float32)[np.newaxis, 1:, np.newaxis]
load = np.repeat(load, len(data_agg), 0)

train_coord, train_load, train_tgt = data_agg[train_index, :, :, :3], load[train_index, :, :], data_agg[train_index, :, :, 6:8]
valid_coord, valid_load, valid_tgt = data_agg[valid_index, :, :, :3], load[valid_index, :, :], data_agg[valid_index, :, :, 6:8]
test_coord,  test_load,  test_tgt  = data_agg[test_index,  :, :, :3], load[test_index,  :, :], data_agg[test_index,  :, :, 6:8]

train_coord, train_load, train_tgt = train_coord.reshape((-1, min_point_num, 3)), train_load.reshape((-1, 1)), train_tgt.reshape((-1, min_point_num, 2))
valid_coord, valid_load, valid_tgt = valid_coord.reshape((-1, min_point_num, 3)), valid_load.reshape((-1, 1)), valid_tgt.reshape((-1, min_point_num, 2))
test_coord,  test_load,  test_tgt  =  test_coord.reshape((-1, min_point_num, 3)),  test_load.reshape((-1, 1)),  test_tgt.reshape((-1, min_point_num, 2))

print(train_coord.shape, train_load.shape, train_tgt.shape)
print(valid_coord.shape, valid_load.shape, valid_tgt.shape)
print( test_coord.shape,  test_load.shape,  test_tgt.shape)

# normalization
coord_max, coord_min = np.max(train_coord, axis=(0, 1)), np.min(train_coord, axis=(0, 1))
load_max,  load_min  = np.max(train_load, axis=0),       np.min(train_load, axis=0)

train_coord, train_load = (train_coord - coord_min) / (coord_max - coord_min), (train_load - load_min) / (load_max - load_min)
valid_coord, valid_load = (valid_coord - coord_min) / (coord_max - coord_min), (valid_load - load_min) / (load_max - load_min)
test_coord,  test_load  = (test_coord  - coord_min) / (coord_max - coord_min), (test_load  - load_min) / (load_max - load_min)

print(np.max(train_coord), np.min(train_coord), np.max(train_load), np.min(train_load), np.max(train_tgt), np.min(train_tgt))
print(np.max(valid_coord), np.min(valid_coord), np.max(valid_load), np.min(valid_load), np.max(valid_tgt), np.min(valid_tgt))
print(np.max(test_coord),  np.min(test_coord),  np.max(test_load),  np.min(test_load),  np.max(test_tgt),  np.min(test_tgt))

# Dataset
train_load_tensor  = torch.from_numpy(train_load).contiguous()
train_coord_tensor = torch.from_numpy(train_coord).contiguous()
train_tgt_tensor   = torch.from_numpy(train_tgt).contiguous()

valid_load_tensor  = torch.from_numpy(valid_load).contiguous()
valid_coord_tensor = torch.from_numpy(valid_coord).contiguous()
valid_tgt_tensor   = torch.from_numpy(valid_tgt).contiguous()

test_load_tensor   = torch.from_numpy(test_load).contiguous()
test_coord_tensor  = torch.from_numpy(test_coord).contiguous()
test_tgt_tensor    = torch.from_numpy(test_tgt).contiguous()

print("train: ", train_coord_tensor.shape, train_load_tensor.shape, train_tgt_tensor.shape)
print("valid: ", valid_coord_tensor.shape, valid_load_tensor.shape, valid_tgt_tensor.shape)
print("test : ", test_coord_tensor.shape, test_load_tensor.shape, test_tgt_tensor.shape)

train_set = TensorDataset(train_coord_tensor, train_load_tensor, train_tgt_tensor)
valid_set = TensorDataset(valid_coord_tensor, valid_load_tensor, valid_tgt_tensor)
test_set  = TensorDataset(test_coord_tensor,  test_load_tensor,  test_tgt_tensor)

train_loader = DataLoader(train_set, batch_size=16)
valid_loader = DataLoader(valid_set, batch_size=1)
test_loader  = DataLoader(test_set,  batch_size=1)
